doc/reference/trd106-completion-codes.md

```diff
@@ -0,0 +1,57 @@
+Application Completion Codes
+========================================
+
+**TRD:** 106 <br/>
+**Working Group:** Kernel<br/>
+**Type:** Documentary<br/>
+**Status:** Draft<br/>
+**Author:** Alyssa Haroldsen<br/>
+**Draft-Created:** December 6, 2021<br/>
+**Draft-Modified:** December 6, 2021<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** tock-dev@googlegroups.com</br>
+
+Abstract
+-------------------------------
+This document describes the expected behavior of application completion codes
+when terminating via the `exit` syscall, as described in [TRD 104][exit-syscall].
+
+1 Introduction
+===============================
+When an application exits via the [`exit` syscall][exit-syscall], it can specify
+a **completion code**, an unsigned 32-bit number which indicates status. This
+information can be stored in the kernel and used in management or policy
+decisions.
+
+This number is called an "exit status", "exit code", or "result code" on other
+platforms.
+
+2 Design Considerations
+===============================
+When possible, Tock should follow existing conventions and terminology from
+other major platforms. This assists in helping the project be more
+understandable to newcomers by following the principle of least astonishment.
+
+3 Design
+===============================
+An application exiting normally via the `exit` syscall MUST use a completion
+code of `0` to indicate normal termination. A non-zero completion code SHOULD be
+used to indicate abnormal termination. A non-zero completion code MAY be the 
+same value as one of the error codes specified in [TRD 104][error-codes].
```

Comment:
It's probably worth making a table here. Something like:

Completion Code | Meaning
------------- | --------
0              | Success
1 - 1024   | Error code as defined in TRD 104
everything else | Not yet defined

Reply:
Notice the MAY - I specifically don't want to reserve 1-1024. The only required convention is 0 vs. non-zero.

Reply:
I added a table that I think indicates this

Reply:
IMO, it seems a little confusing to use MAY here. Either we should reserve 1-1024, or we should just say that 1-MAX is not yet defined. Specifying that it may be defined as an error code, or may not, does not seem to have much value

Reply:
How would you feel about SHOULD? The MAY is to indicate that using 1-1024 for TRD 104 error codes is a good baseline, but an application not following it is still well-formed. I'm thinking from the perspective of an engineer trying to figure out what completion code 12 is, reading this doc, and seeing that it indicates the error code UNINSTALLED by default, providing a valuable hint.

Reply:
I am happy with SHOULD

Reply:
Thoughts on the new language? I also changed the first sentence, switching the logic from "success implies a 0 completion code" to "0 completion code implies success".

Reply:
I'm happy with the new language + table.